[mlir][VectorOps] Add unrolling for n-D vector.interleave ops (3/4)

MacDue opened the pull request:
This unrolls n-D vector.interleave ops like:

```mlir
vector.interleave %i, %j : vector<6x3xf32>
```

To a sequence of 1-D operations:
```mlir
%i_0 = vector.extract %i[0] 
%j_0 = vector.extract %j[0] 
%res_0 = vector.interleave %i_0, %j_0 : vector<3xf32>
vector.insert %res_0, %result[0] :
// ... repeated x6
```

The 1-D operations can then be directly lowered to LLVM.

Depends on: #80966

MacDue committed:
[mlir][VectorOps] Add unrolling for n-D vector.interleave ops

This unrolls n-D vector.interleave ops like:

```mlir
vector.interleave %i, %j : vector<6x3xf32>
```

To a sequence of 1-D operations, which can then be directly lowered to
LLVM.

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorInterleave.cpp:
```diff
@@ -0,0 +1,64 @@
+using namespace mlir::vector;
+
+namespace {
+/// Progressive lowering of InterleaveOp.
```
A bit more doc?

dcaballe commented on mlir/lib/Dialect/Vector/Transforms/LowerVectorInterleave.cpp:
```diff
@@ -0,0 +1,64 @@
+    auto loc = op.getLoc();
+    Value result = rewriter.create<arith::ConstantOp>(
+        loc, resultType, rewriter.getZeroAttr(resultType));
+    for (int idx = 0, end = resultType.getDimSize(0); idx < end; ++idx) {
+      Value extractLhs = rewriter.create<ExtractOp>(loc, op.getLhs(), idx);
+      Value extractRhs = rewriter.create<ExtractOp>(loc, op.getRhs(), idx);
+      Value interleave =
+          rewriter.create<InterleaveOp>(loc, extractLhs, extractRhs);
+      result = rewriter.create<InsertOp>(loc, interleave, result, idx);
+    }
+
+    rewriter.replaceOp(op, result);
```
Could we make this rewrite iterative instead of recursive?

MacDue replied:
It's naturally a recursive problem, so I feel like that'd complicate the rewrite (this is how all other lowerings like this hard handled too).

dcaballe replied:
Note that every time we unroll one dimension the op will go through all the canonicalization/folders the rewrite applies, which is expensive. Given that `ExtractOp` supports multiple indices, wouldn't it be possible to just go over all the dimensions we have to unroll? What am I missing? I'm actually surprised we don't have a utility to do something like that already.

MacDue replied:
AFAIK there's no helper for unrolling vector types (correctly accounting for scalable dimensions). There's the stuff to build such a helper, so I've now added `vector::createUnrollIterator(vType, targetRank)` (to `VectorUtils.h`) and used that for this patch. 

But this is not the normal way these rewrites seem to work (for whatever reason). For example, the lowerings for both `vector.create_mask` and `vector.broadcast` just unroll a single dimension at a time.

dcaballe replied:
Sorry, I didn't intend to enforce this but the codebase is evolving quickly and I think this is one of those things we may want to challenge even if it was done differently somewhere else. Thanks for taking care of it.

c-rhodes commented on mlir/test/Conversion/VectorToLLVM/vector-to-llvm.mlir:
is this testing an intra-dialect conversion with an inter-dialect conversion because it's only integrated in VectorToLLVM? I think the tests would be simpler at Vector dialect level.

MacDue replied:
I'd need to hook it up in the transform dialect to test the lowering alone

MacDue replied:
Done this done :+1:

c-rhodes reviewed (approved):
LGTM cheers

banach-space reviewed (approved):
LGTM, thanks!

MacDue committed:
Fixup: Create and use `vector::createUnrollIterator()` until

Instead of progressively unrolling a leading dimension at a time, this
now uses `vector::createUnrollIterator()` which returns an iterator for
all leading dimensions of a vector type (until a target rank).

hanhanW reviewed (approved):
LGTM, thanks!
